### ISSUE TITLE
MQTT source deadlocks after TCP connection is reset

### ISSUE THREAD

**Issue by nenad** [redpanda-data/connect#717]
Hey Ashley,

I've come across a problem with Benthos which uses the eclipse/paho.mqtt.golang package with versions higher than v1.2.0.
I have two different applications deployed with different Benthos versions, one is v3.29.0 and the other v3.42.1. Both of them use the same MQTT broker as source.
The broker I connect to occasionally resets the connection and I saw that Benthos handles the reconnection internally. However, the older version manages to reconnect after it has logged that TCP error has occurred:
```json
{"@timestamp":"2021-03-31T04:19:39Z","@service":"p2l-bridge","level":"ERROR","component":"p2l-bridge.input","message":"Connection lost due to: write tcp 172.23.38.58:38332->88.99.71.230:8883: write: connection reset by peer"}
{"@timestamp":"2021-03-31T04:19:39Z","@service":"p2l-bridge","level":"INFO","component":"p2l-bridge.input","message":"Receiving MQTT messages from topics: [HF/+/+/+/+/+/+/Induct HF/+/+/+/+/+/Induct]"}
```
However on the newer version this is not the case. The error never even gets logged by Benthos.

I've deployed v3.38.0 which is the latest version with the v1.2.0 of the MQTT library and that one performs the same as the v3.29.0 and successfully recognizes that the connection failed and reconnects.

As I wanted to use the latest and greatest, I compiled Benthos v3.42.1 with debug logs from the MQTT library. Here are the logs from two runs when it inevitably got stuck:

<details><summary>mqtt_stuck_1.txt</summary>
<p>

```
[store]    memorystore del: message 1008 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1012
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1009 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1013
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      startIncoming Received Message
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1010 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1014
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1011 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1015
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1012 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1016
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing oboundp reporting error  write tcp 172.23.86.191:53014->55.55.55.55:8883: write: connection reset by peer
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1013 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1017
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing oboundp reporting error  write tcp 172.23.86.191:53014->55.55.55.55:8883: write: connection reset by peer
[net]      putting puback msg on obound
[store]    memorystore del: message 1014 was deleted
[net]      done putting puback msg on obound
[client]   Connect comms goroutine - error triggered write tcp 172.23.86.191:53014->55.55.55.55:8883: write: connection reset by peer
[client]   internalConnLost called
[client]   stopCommsWorkers called
[net]      outgoing waiting for an outbound message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing oboundp reporting error  tls: use of closed connection
[pinger]   keepalive stopped
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1018
[net]      startIncoming Received Message
[client]   internalConnLost waiting on workers
[client]   stopCommsWorkers waiting for workers
[net]      putting puback msg on obound
[store]    memorystore del: message 1015 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1019
[net]      startIncoming Received Message
[net]      putting puback msg on obound
[store]    memorystore del: message 1016 was deleted
```

</p>
</details>

<details><summary>mqtt_stuck_2.txt</summary>
<p>

```
[net]      startIncoming Received Message
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 697 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 701
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      startIncoming Received Message
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 698 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 702
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 699 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      startIncoming Received Message
[net]      startIncomingComms: received publish, msgId: 703
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 700 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 704
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 701 was deleted
[net]      done putting puback msg on obound
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      logic waiting for msg on ibound
[net]      outgoing waiting for an outbound message
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 705
[net]      startIncoming Received Message
[net]      putting puback msg on obound
[store]    memorystore del: message 702 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      startIncomingComms: received publish, msgId: 706
[net]      startIncoming Received Message
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 703 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 707
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 704 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      startIncoming Received Message
[net]      startIncomingComms: received publish, msgId: 708
[net]      outgoing oboundp reporting error  write tcp 172.23.84.39:37480->55.55.55.55:8883: write: connection reset by peer
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 705 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 709
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing oboundp reporting error  write tcp 172.23.84.39:37480->55.55.55.55:8883: write: connection reset by peer
[net]      putting puback msg on obound
[store]    memorystore del: message 706 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 710
[net]      startIncoming Received Message
[net]      putting puback msg on obound
[store]    memorystore del: message 707 was deleted
[pinger]   ping check 3.294454048
[pinger]   ping check 8.29445355
[pinger]   ping check 13.294459073
[pinger]   ping check 18.294462736
[pinger]   ping check 23.294453457
[pinger]   ping check 28.294450087
[pinger]   ping check 33.294453784
[pinger]   keepalive sending ping
[pinger]   write tcp 172.23.84.39:37480->55.55.55.55:8883: write: connection reset by peer
[pinger]   ping check 4.999926672
[pinger]   ping check 9.999940218
[pinger]   ping check 14.999981227
[client]   internalConnLost called
[client]   stopCommsWorkers called
[client]   internalConnLost waiting on workers
[client]   stopCommsWorkers waiting for workers
```

</p>
</details>

Both runs have these lines in common which is the blocking part I believe:
```
[client]   internalConnLost waiting on workers
[client]   stopCommsWorkers waiting for workers
```

However, the error `write tcp 172.23.84.39:37480->55.55.55.55:8883: write: connection reset by peer` was never reported by Benthos like in the older versions. I don't know if this is Benthos not handling the error correctly, or the library not propagating the error upwards.

Additionally, I've also enabled the debug endpoints and ran all debug tools on the application when it got stuck. They are attached to this issue.
[symbol.txt](https://github.com/Jeffail/benthos/files/6234587/symbol.txt)
[stack.txt](https://github.com/Jeffail/benthos/files/6234588/stack.txt)
[debug.zip](https://github.com/Jeffail/benthos/files/6234594/debug.zip)

At the moment to workaround the issue I'm using v3.38.0.

Thanks for looking into this!

**Comment by Jeffail** [comment]
Hey @nenad, thanks for the great write up! There's some ugly work arounds on the Benthos side for recovering lost connections, but then after introducing that we upgraded to 1.3.1 of the library which likely fixed the issue we were working around (or caused the work around to break).

I'll try and reproduce the deadlock locally with the latest build, then I'll remove the ugly hacks around stale connections and configure the library itself to auto-reconnect, I'll also upgrade to the latest (currently 1.3.2) as that seems to include a fix that might be relevant.

**Comment by nenad** [comment]
Just tested again by compiling Benthos from master and updating the paho.mqtt.golang version to v1.3.2 and the issue is still there.

Yesterday I tried to reproduce the same issue locally by using `socat` to forward the TCP connection to a local broker, but I could not manage to enter in the deadlock state. Just now I realized that I was testing with very low load (1 msg/min), but on production I receive around 200 msg/min.

I'll give it another run by turning auto-reconnecting to on now.

**Comment by Jeffail** [comment]
Relevant prior issue: https://github.com/Jeffail/benthos/issues/479

@nenad, the last time we were stung with the client lib silently losing connections I added a field `stale_connection_timeout`, that's the nuclear option if the auto reconnect doesn't work, more details here: https://github.com/Jeffail/benthos/issues/479#issuecomment-662310326

**Comment by nenad** [comment]
So, this is what I tested now:
- Auto reconnect set to `true`
- Manual reconnect logic was reverted (introduced in 3c7fce2c74d00453619997eef13f46054526c1d4)

However, the problem still persists. This time I didn't even get a log out of the MQTT library saying that the TCP connection was broken and it still got stuck. I checked with `netstat` in the container that I'm running Benthos if we have a connection to the MQTT broker and there wasn't one.

<details>
<summary>Logs from MQTT lib</summary>
<p>

```
[net]      startIncoming Received Message
[net]      outgoing waiting for an outbound message
[net]      startIncomingComms: received publish, msgId: 1131
[net]      putting puback msg on obound
[store]    memorystore del: message 1128 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1132
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      startIncoming Received Message
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1129 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1133
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1130 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1134
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1131 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1135
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1132 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1136
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1133 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1137
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      outgoing waiting for an outbound message
[net]      putting puback msg on obound
[store]    memorystore del: message 1134 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1138
[net]      startIncoming Received Message
[net]      obound priority msg to write, type *packets.PubackPacket
[net]      putting puback msg on obound
[store]    memorystore del: message 1135 was deleted
[net]      done putting puback msg on obound
[net]      logic waiting for msg on ibound
[net]      startIncomingComms: got msg on ibound
[net]      startIncomingComms: received publish, msgId: 1139
[net]      startIncoming Received Message
[net]      putting puback msg on obound
[store]    memorystore del: message 1136 was deleted
[pinger]   ping check 0.970677193
[pinger]   ping check 5.970670817
[pinger]   ping check 10.970672776
[pinger]   ping check 15.970675942
[pinger]   ping check 20.970671871
[pinger]   ping check 25.970668925
[pinger]   ping check 30.97061431
[pinger]   keepalive sending ping
[pinger]   ping check 5.000006255
[pinger]   ping check 10.000010252
[client]   internalConnLost called
[client]   stopCommsWorkers called
[client]   internalConnLost waiting on workers
[client]   stopCommsWorkers waiting for workers
```
</p>
</details>

This makes me suspect that the library has a deadlock issue introduced in v1.3.0 which was a release that had lots of refactoring work done.

**Comment by Jeffail** [comment]
Thanks @nenad, I've marked this as "waiting for upstream" since it's looking like we'll need a fix in the client library. If you're able to reproduce this with a minimal example then it would be worth raising it over at https://github.com/eclipse/paho.mqtt.golang, but ideally we need to have an example that rules out any chance of blocking in the callbacks.

An option in the meantime would be to roll back to v1.2.0, but that also has the possibility of re-introducing old bugs. I'll mull it over for a little bit.

**Comment by twmb** [comment]
Closing — the underlying paho.mqtt.golang library has been upgraded from v1.3.x (where this deadlock was observed) to v1.5.1, which has had significant refactoring and bug fixes since. If this issue persists with the current version, please open a new issue.
